Crangler

kroening opened the pull request:
This adds a new command-line tool, which replaces current flows that enact changes to the program that is analysed. The new flow is to make these changes before parsing, which enables more far-reaching changes and changes that depend on scoped identifiers.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
Add the crangler tool

This is a command-line utility that makes changes to a preprocessed C file
that are prescribed in a JSON configuration file.

The initial set of transformations is as follows.
 * add a contract (pre/post/assigns) to a named C function
 * add a loop invariant to a loop identified by the name
   of the function its in and a loop number
 * remove the 'static' storage classifier from a function

The resulting source file is written to standard output or to file
specified in the JSON configuration.

tautschnig commented:
Two high-level comments:
1) Do we really need YAML? Couldn't we use JSON instead and have people use tools like `yq` to convert?
2) Even if not a particularly elaborate regression test, could we at least have an example of an application?

tautschnig commented on src/crangler/crangler_parse_options.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+  if(cmdline.args.empty())
+  {
+    std::cerr << "please give a configuration file\n";
```
Use error logging via the message handler instead?

tautschnig commented on src/crangler/Makefile:
```diff
@@ -0,0 +1,36 @@
+SRC = c_defines.cpp \
```
Somewhat arbitrary place for comment: we're missing a CMake configuration for this new tool.

tautschnig commented on src/crangler/c_defines.cpp:
```diff
@@ -0,0 +1,67 @@
+    if(space_pos == std::string::npos)
+      continue;
+
+    auto id = line.substr(8, space_pos - 8);
+    auto value = line.substr(space_pos + 1, std::string::npos);
+    map[id].value = value;
```
Do we need to care about `#undef` and a possible re-definition?

tautschnig commented on src/crangler/c_defines.cpp:
```diff
@@ -0,0 +1,67 @@
+
+  // output
+  std::ostringstream out;
+  for(auto &t : tokens)
+  {
+    if(is_identifier(t))
+    {
+      auto m_it = map.find(t.text);
+      if(m_it != map.end())
+      {
+        out << m_it->second.value;
+      }
+      else
+        out << t.text;
+    }
+    else
+      out << t.text;
+  }
```
What about macros that use other macros? What about macros that take parameters (see `getc_unlocked(fp)` example in the comment)?

tautschnig commented on src/crangler/c_defines.h:
```diff
@@ -0,0 +1,37 @@
+public:
+  struct definet
+  {
+    optionalt<std::vector<std::string>> parameters;
```
Never used (although they'd need to be taken into account).

tautschnig commented on src/crangler/c_defines.h:
```diff
@@ -0,0 +1,37 @@
+#include <unordered_map>
+#include <vector>
+
+class c_definest
```
A brief documentation of the purpose of this class would be useful.

kroening replied:
added a comment

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+
+void c_wranglert::configure_sources(const jsont &config)
+{
+  // look for the 'sources' key
+  auto sources = config["sources"];
```
There aren't a lot of comments in this code, but then this one stating the obvious. Perhaps I'm missing something and it's not actually doing the obvious?

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+    }
+  }
+
+  // look for the 'includes' key
```
See above.

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+      }
+      else if(split[0] == "stub")
+      {
+        std::string rest;
+        for(std::size_t i = 1; i < split.size(); i++)
+        {
+          if(i != 1)
+            rest += ' ';
+          rest += split[i];
+        }
+
```
Use `join_strings`?

kroening replied:
done

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+        (split[0] == "for" && split.size() >= 3 && split[2] == "invariant") ||
+        (split[0] == "while" && split.size() >= 3 && split[2] == "invariant"))
+      {
+        std::string rest;
+        for(std::size_t i = 3; i < split.size(); i++)
+        {
+          if(i != 3)
+            rest += ' ';
+          rest += split[i];
+        }
```
Use `join_strings`?

kroening replied:
done

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+      }
+      else if(split[0] == "assert" && split.size() >= 3)
+      {
+        std::string rest;
+        for(std::size_t i = 2; i < split.size(); i++)
+        {
+          if(i != 2)
+            rest += ' ';
+          rest += split[i];
+        }
```
Use `join_strings`?

kroening replied:
done

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+        split[0] == "ensures" || split[0] == "requires" ||
+        split[0] == "assigns")
+      {
+        std::string rest;
+        for(std::size_t i = 1; i < split.size(); i++)
+        {
+          if(i != 1)
+            rest += ' ';
+          rest += split[i];
+        }
```
Use `join_strings`?

kroening replied:
done

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+    return;
+
+  if(!output.is_string())
+    throw deserialization_exceptiont("outputs entry must be string");
```
Perhaps "output" instead of "outputs?"

kroening replied:
done

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+std::string
+preprocess(const std::string &source_file, const c_wranglert &c_wrangler)
+{
+  std::vector<std::string> argv = {"cc", "-E", source_file};
+
+  for(const auto &include : c_wrangler.includes)
+  {
+    argv.push_back("-I");
+    argv.push_back(include);
+  }
+
+  for(const auto &define : c_wrangler.defines)
+    argv.push_back(std::string("-D") + define);
+
+  std::ostringstream result;
+
+  auto run_result = run("cc", argv, "", result, "");
+  if(run_result != 0)
+    throw system_exceptiont("preprocessing " + source_file + " has failed");
+
+  return result.str();
```
Should we have a Windows/MSVC version thereof?

tautschnig commented on src/crangler/c_wrangler.cpp:
```diff
@@ -0,0 +1,474 @@
+c_definest
+get_defines(const std::string &source_file, const c_wranglert &config)
+{
+  std::vector<std::string> argv = {"cc", "-E", "-dM", source_file};
+
+  for(const auto &include : config.includes)
+  {
+    argv.push_back("-I");
+    argv.push_back(include);
+  }
+
+  std::ostringstream result;
+
+  auto run_result = run("cc", argv, "", result, "");
+  if(run_result != 0)
+    throw system_exceptiont("preprocessing " + source_file + " has failed");
+
+  c_definest defines;
+  defines.parse(result.str());
+  return defines;
```
Windows/MSVC?

tautschnig commented on src/crangler/c_wrangler.h:
```diff
@@ -0,0 +1,19 @@
+#ifndef CPROVER_CRANGLER_C_WRANGLER_H
+#define CPROVER_CRANGLER_C_WRANGLER_H
+
+#include <util/json.h>
```
`class jsont;` might do?

kroening replied:
done

tautschnig commented on src/crangler/mini_c_parser.cpp:
```diff
@@ -0,0 +1,373 @@
+
+  while(!eof())
+    result.push_back(parse_declaration());
+  // std::cout << "TOKEN: " << consume_token() << "\n";
```
Remove?

kroening replied:
done

tautschnig reviewed (approved):
Approving after some further out-of-band discussion. In addition to my earlier comments: some further tests and/or documentation would be really helpful.

tautschnig committed:
Crangler: support removing static from objects

Just like functions, objects can be file-local. To support accessing
them, permit removing "static" from objects as well.

tautschnig committed:
Crangler: support regular expressions to select functions

To avoid having to list large numbers of functions for mangling, permit
regular expressions as function name entries in the JSON configuration
file.

tautschnig commented:
We might still want to add a capability to perform name mangling, but once done we should be able to get rid of the `export-file-local-symbols` cross-cutting changes.

TGWDB reviewed (approved):
This is an approval for the Cmake/make aspects as these were blocked by OpenSource/Codeowner
